WIP: Implement search results pagination

felixSchl opened the pull request:
Stab at #305

It's working but the HTML is pretty horrifying for large result sets as there's a link to every single page in the navigation bar. I was thinking this would be a nice way to render the navigation: https://stackoverflow.com/a/14193775/2954634?

Also, in order to know the total amount of pages, I removed the [`take 50`](https://github.com/purescript/pursuit/blob/c16211e916b52f446d9f7ae4dce967629ffac0d5/src/Handler/Search.hs#L90), so it's doing a lot more work now in case of large result sets. If so desired, we could simply take 51, then check if the result is greater than 50, then simply render a "next" link, as opposed to individual page links.

Would be great to get some feedback on the above so I can adjust the code accordingly

hdgarrood commented:
In my mind it's a lot more important that the load on the server is kept down than the user knows how many pages there are; from that perspective I don't think we can justify calculating the total amount of pages immediately. I can't imagine a context in which it would be useful to know how many pages of results there are.

How about the first page displays the first 50 results, as it does currently, but then, if there are more results, there is a 'next page' link which takes you to page 2, and then if there are still more, there is a 'next page' link for page 3, and so on?

I think we should still have a cutoff point where the server refuses to do any more work to find results. I think 10 pages is probably a reasonable cutoff point -  if there are still more results after that I think it's safe to say it's unlikely they're going to be useful, and so instead of rendering a 'next page' link, we could have a message that indicates that there are more results which aren't shown.

felixSchl commented:
I've applied part of the suggestion, but the problem that I see is that the server still has to do a lot of work when navigating to say page 20 since it needs to `drop` from the result set. I guess we could cache search results in memory for a certain amount of time since last access? But I feel this introduces non-essential complexity.

hdgarrood commented:
Yes, I don't want to do caching like that, I agree that that could be very complex and not really give us all that much in return. That's why I proposed having hard cut-off point at 10 pages.

felixSchl commented:
Alright, I have added such a limit

hdgarrood commented on src/Handler/Search.hs:
```diff
@@ -29,17 +28,20 @@ import qualified XMLArrows
       (results, hasMore) <- case tryParseType query of
-        Just ty | not (isSimpleType ty) -> searchForType page ty
-        _ -> searchForName page (toLower query)
+        Just ty | not (isSimpleType ty) -> searchForType (min maxPages page) ty
```
I think we should extract the part that makes sure `page <= maxPages` into a separate function at the time we take it out of the query string, together with all of the other validations and things we do to it.

hdgarrood commented on src/Handler/Search.hs:
```diff
@@ -7,6 +7,10 @@ module Handler.Search
 import Import
 import Data.Trie (elems, submap)
 import Data.Version (showVersion)
+import Data.Maybe (isNothing)
```
Can you re-export these functions and `MaybeT` from `Import.NoFoundation`? I think they're ubiquitous enough that it's better to do that to help avoid having too many imports here.

hdgarrood commented on src/Handler/Search.hs:
```diff
@@ -21,22 +25,52 @@ import qualified Language.PureScript as P
 getSearchR :: Handler TypedContent
 getSearchR = do
   mquery <- lookupGetParam "q"
+  page <- maybe 1 (max 1) . ((readMaybe . unpack) =<<) <$> lookupGetParam "page"
```
I think we should extract all of the parsing and validation logic for the page parameter into a separate function.

hdgarrood commented on src/Handler/Search.hs:
```diff
@@ -21,22 +25,52 @@ import qualified Language.PureScript as P
+      MaybeT $ do
+        urp <- getUrlRenderParams
+        getParams <- Map.fromList . reqGetParams <$> getRequest
+        let mkPageLink p = urp cr $ Map.toList
```
For the pagination widget, how about having a link for all pages up to the current one? Also can you explain what the purpose of `getParams` is here?

felixSchl replied:
So you mean the pages keep growing as you advance? So on page 4 you would see "1 2 3 [4] next", but on page 3 you would only see "1 2 [3] next"?

hdgarrood replied:
Yes - the idea being, you might think "okay I should have seen the thing I am looking for by now so I don't think looking further will help, but I remember seeing something on page 2 which was close to what I want so maybe the answer is in there somewhere"

hdgarrood replied:
Also because if we're on page 4, we know that pages 1, 2, and 3 exist, but we don't know if 6,7, or 8 exist.

felixSchl replied:
Sorry, forgot to explain `getParams`, it's to modify the `?page=` param for each link

hdgarrood commented on src/Handler/Search.hs:
```diff
@@ -84,15 +118,24 @@ routeResult SearchResult{..} =
+searchForName page query = do
+  db <- atomically . readTVar =<< (appDatabase <$> getYesod)
+  let xs = map fst
+            $ take (resultsPerPage + 1)
```
This new logic seems like it should be extracted into a separate function too.

felixSchl commented:
Alright, I've addressed all of the above

hdgarrood commented on src/Handler/Search.hs:
```diff
@@ -52,6 +85,9 @@ getSearchR = do
     isSimpleType _ = False
 
+    lookupCurrentPage :: Handler Int
+    lookupCurrentPage = maybe 1 (max 1) . ((readMaybe . unpack) =<<) <$> lookupGetParam "page"
```
Could  `(min maxPages page)` go in here too?

felixSchl replied:
Sure. I've rewritten it in a more verbose but more readable way.

hdgarrood replied:
:+1:

hdgarrood commented on src/Handler/Search.hs:
```diff
@@ -84,15 +120,21 @@ routeResult SearchResult{..} =
-searchForName query = do
+applyPagination :: Int -> [a] -> ([a], Bool)
+applyPagination page xs =
+    let ys = take (resultsPerPage + 1) $ drop ((page - 1) * resultsPerPage) xs
```
This will give us `resultsPerPage + 1` results per page, right? That doesn't seem right :smile:

How about this?

```haskell
let
  (ys, more) = splitAt resultsPerPage $ drop ((page - 1) * resultsPerPage) xs
in
  (ys, not (null more))
```

hdgarrood replied:
Also this function could do with a brief comment, I think

felixSchl replied:
nice

hdgarrood commented on src/Handler/Search.hs:
```diff
@@ -84,15 +120,21 @@ routeResult SearchResult{..} =
+searchForType page ty = do
   db <- atomically . readTVar =<< (appDatabase <$> getYesod)
-  return (map fst (take 50 (concat (elems (submap (encodeUtf8 query) db)))))
+  return $ first (map fst) $ applyPagination page $ sortBy (comparing snd) (mapMaybe (matches ty) $ concat (elems db))
```
This line is getting a bit too long now, I think

hdgarrood commented:
Also yes, let's use a header in the JSON response to indicate whether there are more results.

hdgarrood commented on src/Handler/Search.hs:
```diff
@@ -86,7 +86,12 @@ getSearchR = do
     lookupCurrentPage :: Handler Int
-    lookupCurrentPage = maybe 1 (max 1) . ((readMaybe . unpack) =<<) <$> lookupGetParam "page"
+    lookupCurrentPage = fmap (fromMaybe 1) <$>
+        runMaybeT $ do
```
I think this would be slightly clearer without `MaybeT`, actually:

```haskell
do
  mpage <- lookupGetParam "page"
  return $ case mpage >>= (readMaybe . unpack) of
    Just page -> max 1 $ min maxPages page
    Nothing -> 1

felixSchl commented:
Alright, I've removed the use of `MaybeT` and am now rendering the `Link` header (I took inspiration from: https://developer.github.com/v3/guides/traversing-with-pagination/#navigating-through-the-pages). Sorry for the amount of changed code in latest commit, but there was a bit of common logic between rendering the header and the widget, so I extracted

hdgarrood commented on src/Handler/Search.hs:
```diff
@@ -21,26 +24,87 @@ import qualified Language.PureScript as P
+        getParams <- Map.fromList . reqGetParams <$> getRequest
+        return $ Just $ \p ->
+          urp cr $ Map.toList
+                 $ Map.insert "page" (pack $ show p) getParams
```
Come to think of it I'm not sure `MaybeT` is really adding much here either.

hdgarrood commented on static/css/pursuit.css:
```diff
@@ -619,6 +619,47 @@ ol li {
+}
+
+.pagination {
+    position: relative;
```
Why are the `position` rules needed?

felixSchl replied:
I am no CSS wiz, I'll try to remove

felixSchl commented on src/Handler/Search.hs:
```diff
@@ -21,26 +24,87 @@ import qualified Language.PureScript as P
+        Just mkPageLink ->
+          let allPages = [1..page]
+              mmNextPageLink = if hasMore
+                                  then Just $ if page >= maxPages
```
I think for a one off job like this the cognitive burden of `Maybe (Maybe a)` is less than introducing another ADT the reader of the code has to keep in memory (and then forget about after realizing how irrelevant it is)

hdgarrood replied:
In the past I would have agreed with you, but every single time I've used `Maybe (Maybe a)` like this, I've regretted it. In my opinion it's a far larger cognitive burden to have to remember what the meaning of `Nothing` vs `Just Nothing` is.

However before you address that, I noticed that this is what page 10 now looks like on my machine:
![screenshot from 2017-07-15 13-28-42](https://user-images.githubusercontent.com/1270186/28239460-a62e31d4-6964-11e7-9c4c-789764335141.png)

and I was thinking maybe it would make more sense to have the "further results have been omitted" message after the last search result rather than inside the pagination widget.

Also I think the clickable regions in the "goto page n" links should still be made a little larger. Also I don't think there should be non-clickable gaps in between the links. It might also be nice to have the background change colour slightly on hover so that it's clear what you're clicking on. Finally I think it might be slightly nicer to have the next page link text include a chevron character like "next page &raquo;" and similar for the prev page link?

felixSchl commented on src/Handler/Search.hs:
```diff
@@ -21,26 +24,87 @@ import qualified Language.PureScript as P
+    jsonOutput :: [SearchResult] -> Int -> Bool -> Handler Value
+    jsonOutput results page hasMore = do
+        when (hasMore && page <= maxPages) $ void $ do
+          mkPaginationHeader page hasMore >>= \mHeader -> case mHeader of
```
I didn't want to add any syntax extensions, but I think `LambdaCase` would be pretty cool for all these cases:

```
mkPaginationHeader page hasMore >>= case
  Nothing -> pure ()
  Just ...
```

felixSchl replied:
Oh, hold on, it's already enabled, will transform

felixSchl commented:
Alright, I've addressed all the things

hdgarrood commented on templates/search.hamlet:
```diff
@@ -44,3 +48,12 @@
+  $if hasOmittedResults
+    <hr />
+    <div .result__omission>
+      More results have been omitteded
```
I think "Further results have been omitted" would be better here.

hdgarrood commented on static/css/pursuit.css:
```diff
@@ -619,6 +619,59 @@ ol li {
 .result__actions__item + .result__actions__item {
   margin-left: 0.65em;
 }
+.result__omission {
```
I just need to work out how we will deal with this CSS, because it's generated from a LESS file (in this repo), but there's also a copy of it in the PureScript repo, and they're starting to diverge.

hdgarrood replied:
https://github.com/purescript/purescript/issues/3016 if you're wondering what's going on

hdgarrood commented on src/Handler/Search.hs:
```diff
@@ -21,26 +23,87 @@ import qualified Language.PureScript as P
 
 import qualified XMLArrows
 
+resultsPerPage :: Int
```
Was this 50 originally?
